### PR TITLE
Add Ledger API test tool tests for exceptions

### PR DESCRIPTION
**Pull request by cocreature**
Currently testing against:
1. Sandbox classic in-memory
2. Sandbox classic on SQL
3. Ledger on SQL
4. Ledger on Memory

Requests for additional tests are covered in https://github.com/digital-asset/daml/issues/9526 and will be adressed separately.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-on-memory/BUILD.bazel [hunk 1]
```diff
@@ -136,6 +136,12 @@ conformance_test(
         "--contract-id-seeding=testing-weak",
         "--participant=participant-id=example,port=6865",
     ],
+    lf_versions = [
```
Comment (cocreature) [inline ledger/ledger-on-memory/BUILD.bazel:139]: Obviously not going to merge it like this, this is a draft for a reason.

**Review by meiersi-da — COMMENTED** [review]
@cocreature : this goes in a good direction. I believe many of these tests are related to the read-write conflict tests introduced as part of #9138. With exceptions we get a new set of read-write conflicts. Here are the existing ones for reference:

1. archive-contract vs. non-consuming-choice
2. archive-contract vs fetch-contract
3. archive-contract vs. successful lookup-by-key
4. non-transient-create-with-key vs. transient create-with-key
5. non-transient-create-with-key vs. failing lookup-by-key 

I believe we can extend this for exceptions as follows:

1. archive-contract vs. consuming or non-consuming choice within rollback
2. archive-contract vs fetch-contract within rollback
3. archive-contract vs. successful lookup-by-key within rollback
4. non-transient-create-with-key vs. transient or non-transient create-with-key within rollback
5. non-transient-create-with-key vs. failing lookup-by-key within rollback

I'd expect to test these once in a sequential fashion to see that the expected failure happens and once within the race-condition setting to see that the conflict detection logic for concurrent transactions works as expected.

What do you think?

**Comment by cocreature** [comment]
@meiersi-da 
So looking through your list, 2 seems already covered by `ExRollbackActive`. 4 is covered by `ExRollbackDuplicateKey` for the non-transient part, happy to add something for the transient part.
1 and 3 seems very sensible, happy to add those. If I understand you correctly, 5 is just 4 but you start with the failing lookup and then create it instead of starting with the successful lookup and then archiving it.

As for the race condition tests, I agree those are sensible but I’d prefer to tackle it in to steps and get something in that we agree on first for the semantics of the sequential cases and then add the race condition tests on top of that.

**Comment by cocreature** [comment]
- 270e5d3a6bf63033dff0690a823b72cbff7928ae for 5 (or rather key fetches)
- ee47d460ebdd27be0c28e5fb10a166ba304cecdf for 1 (consuming and nonconsuming)
- 1bef91812f9785e8f4addec5799757c97c8022c4 as a reverse version of 4
- 2337a99109cfdcbf98011075b22911503c866788 for nested try/catch as requested by @mziolekda 

What did I forget? And are you fine with leaving the race condition tests for a separate step?

**Comment by meiersi-da** [comment]
@cocreature : LGTM. I agree wrt adding RaceCondition tests in a separate step once the sequential versions work.

**Comment by oggy-** [comment]
A few more test cases that I'd add:
1. liveness tests for non-conflicts:
    a. a rolled-back consuming exercise doesn't block later consuming or non-consuming exercises (you have one of these covered I believe with your archive after consuming exercise, but not for a non-consuming one)
    b. a rolled-back key creation doesn't block a subsequent creation of the same key
2. check that conflicting and non-conflicting (liveness above) work if you have them in the same transaction, with interesting projections and two nodes

**Comment by cocreature** [comment]
@oggy- covered 1 in 50eb70090cc3b0db3383932fba75e254948d5475 (at least I tried :))

As for 2, do you have something specific in mind there? I can add some random stuff but I think I don’t really have an intuition of what is interesting there.

**Comment by oggy-** [comment]
I think all of the following would be useful to test the internal consistency checks of the ledger under test:
1. repeat the existing tests as-is, just wrapped  (insofar possible) in a single transaction.
2. same as 1, but roll the transaction back, and see that the accept/reject behavior doesn't change.
3. repeat 1 and 2, but now intersperse the transactions (including their rolled back parts) with actions that have different informees than the other actions, such that these informees are hosted on a different participant. 

Possibly useful to combine with the existing checks, such that we test whether the ledger computes the inputs/outputs of a transaction correctly given partially rolled-back. Or maybe checking the ACS suffices.

I know I'm asking for a lot of tests :) But it's a complicated feature and I see the potential for errors in all of these contexts. Maybe there's a clever way to remove some of the repetition...

**Comment by cocreature** [comment]
I’m a bit worried this is getting slightly out of hand. The original exceptions doc asked for 5 different test cases, we are now up to 15 + requests to run all of those in 3 different variations (race condition, single transaction, rollback the whole transaction) and other tests.

**Comment by oggy-** [comment]
Sure, like I said I understand that I'm asking for a large number of tests. I'm not asking you to built them in this PR. However, without these additional tests, we have no coverage for internal consistency checks in the presence of rollbacks. And we've had several problems in that area with contract keys for example.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/dummy/ExceptionsIT.scala [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.suites
+
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+
+// Empty dummy suite for LF versions that do not support exceptions.
```
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/dummy/ExceptionsIT.scala:8]: If someone has a nicer suggestion here, I’m all ears. This isn’t super nice but it works and is somewhat extensible to more tests that are version specific.

**Comment by cocreature** [comment]
This now passes on Sandbox classic with the SQL backend. I’ve captured the requests for additional tests in #9526. Given that I’d like it to keep working and test it on more ledgers, I would like to merge it with the current sets of tests and then address those requests separately.

**Comment by cocreature** [comment]
I’ve extended this with more ledgers now that they’re supported.

File: daml-lf/language/daml-lf.bzl [hunk 3]
```diff
@@ -72,3 +72,38 @@ LF_MAJOR_VERSIONS = ["1"]
...
+    if a == "1.dev" and b == "1.dev":
+        return 0
+    elif a == "1.dev":
+        return 1
+    elif b == "1.dev":
+        return -1
+    else:
+        aVer = _to_major_minor(a)
+        bVer = _to_major_minor(b)
+        _cmp(aVer, bVer)
```
Comment (sofiafaro-da) [inline daml-lf/language/daml-lf.bzl:98]:
```suggestion
        return _cmp(aVer, bVer)
```

File: daml-lf/language/daml-lf.bzl [hunk 4]
```diff
@@ -72,3 +72,38 @@ LF_MAJOR_VERSIONS = ["1"]
...
+def _gte(a, b):
+    return _cmp(a, b) >= 0
+
+def _lte(a, b):
+    return _cmp(a, b) <= 0
+
+versions = struct(
+    lte = _lte,
+    gte = _gte,
+)
```
Comment (sofiafaro-da) [inline daml-lf/language/daml-lf.bzl:109]: should these be using `_cmp_lf_version` instead of `_cmp`?

Comment (cocreature) [reply daml-lf/language/daml-lf.bzl]: very good point 96d8b10dc1cbe4164ae1821ffc85b14f69c93b7c

**Review by sofiafaro-da — APPROVED** [review]
This looks good. Thank you! 👍

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ExceptionsIT.scala [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      tExercise <- ledger.create(party, ExceptionTester(party))
+      _ <- ledger.exercise(party, t.exerciseRollbackNonConsuming(_, tExercise))
+      _ <- ledger.exercise(party, t.exerciseArchive(_))
+      failure <- ledger
+        .exercise(party, t.exerciseRollbackNonConsuming(_, tExercise))
+        .mustFail("contract is archived")
+    } yield {
+      assertGrpcError(failure, Status.Code.ABORTED, "Contract could not be found")
+    }
+  })
```
[the remark this answers is not in the document]
Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ExceptionsIT.scala]: 00ee366a4609d3f79f6f19a71a28e32ab05835ae

[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ExceptionsIT.scala]: Thanks! :-)

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/dummy/ExceptionsIT.scala (hunk 2 above)
Comment (gerolf-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/dummy/ExceptionsIT.scala:8]: It might be worth adding a "noop" test case with a test title like "empty test cases because DAML LF under test doesn't support exceptions".

**Review by gerolf-da — APPROVED** [review]
Thank you, @cocreature!

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Tests.scala [hunk 6]
```diff
@@ -7,12 +7,19 @@ import java.nio.file.Path
...
 import com.daml.ledger.api.testtool.suites._
+import com.daml.lf.language.LanguageVersion
+import com.daml.ledger.test.TestDar
 
 import scala.collection.SortedSet
 import scala.concurrent.duration.FiniteDuration
 
 object Tests {
 
+  private val supportsExceptions: Boolean = {
```
Comment (nmarton-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Tests.scala:18]: These are temp code, right? after exceptions table (or preview?) this will be gone, right?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/dummy/ExceptionsIT.scala (hunk 2 above)
Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/dummy/ExceptionsIT.scala:8]: I’m not quite following. Users should never be able to run this. We just need to get this to compile but we don’t add it to the list of available test cases for an LF version that does not support it.

**Review by nmarton-da — COMMENTED** [review]
LGTM
on the exhaustiveness of the ExceptionsIT suit regarding the feature I am not sure I am the best person to judge

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Tests.scala (hunk 6 above)
Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Tests.scala:18]: Not quite, we build the ledger API test tool for very old LF versions and those are never going to support LF versions. Once we stabilize it this will move to a stable LF version. If we ever stop building the ledger APi test tool for old LF versions, we can kill this completely.

File: ledger/ledger-on-sql/BUILD.bazel [hunk 7]
```diff
@@ -245,12 +245,15 @@ da_scala_test_suite(
...
 [
+    # TODO https://github.com/digital-asset/daml/issues/8020
+    # Consider if we want to keep dev once exceptions are out of 1.dev
     conformance_test(
         name = "conformance-test-{}".format(db["name"]),
         lf_versions = [
             "stable",
             "latest",
             "preview",
+            "dev",
```
Comment (nmarton-da) [inline ledger/ledger-on-sql/BUILD.bazel:256]: I see you implemented already the fix for the TransactionCommitter, so this makes sense :+1:

**Review by nmarton-da — APPROVED** [review]
:shipit:

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Tests.scala (hunk 6 above)
Comment (nmarton-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Tests.scala:18]: Oh, well you know what you do, but going forward like this might increase this type of clutter, so eventually we might need something else. for now it's fine by me.

**Comment by nmarton-da** [comment]
> LGTM
> on the exhaustiveness of the ExceptionsIT suit regarding the feature I am not sure I am the best person to judge

just read that Simon and Oggy where involved on that topic :man_facepalming: all good from my side :+1:
